File: Source/OldCode/Archive/2004 TVA Code Library/Source/EE/Phasor/ChannelCellBase.vb
'*******************************************************************************************************
'  ChannelCellBase.vb - Channel frame cell base class
'  Copyright © 2004 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2003
'  Primary Developer: James R Carroll, System Analyst [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  3/7/2005 - James R Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Namespace EE.Phasor

    ' This class represents the common implementation of the protocol independent representation of any kind of data cell.
    Public MustInherit Class ChannelCellBase

        Inherits ChannelBase
        Implements IChannelCell

        Private m_parent As IChannelFrame
        Private m_alignOnDWordBoundry As Boolean

        Protected Sub New(ByVal parent As IChannelFrame, ByVal alignOnDWordBoundry As Boolean)

            MyBase.New()

            m_parent = parent
            m_alignOnDWordBoundry = alignOnDWordBoundry

        End Sub

        Protected Sub New(ByVal parent As IChannelFrame, ByVal alignOnDWordBoundry As Boolean, ByVal state As Object, ByVal binaryImage As Byte(), ByVal startIndex As Integer)

            Me.New(parent, alignOnDWordBoundry)

            ParseHeader(state, binaryImage, startIndex)
            startIndex += HeaderLength
            ParseBody(state, binaryImage, startIndex)

        End Sub

        ' Final dervived classes must expose Public Sub New(ByVal parent As IChannelFrame, ByVal state As ChannelFrameParsingState, ByVal index As Integer, ByVal binaryImage As Byte(), ByVal startIndex As Integer)

        ' Derived classes are expected to expose a Protected Sub New(ByVal channelCell As IChannelCell)
        Protected Sub New(ByVal channelCell As IChannelCell)

            Me.New(channelCell.Parent, channelCell.AlignOnDWordBoundry)

        End Sub

        Protected MustOverride Sub ParseHeader(ByVal state As Object, ByVal binaryImage As Byte(), ByVal startIndex As Integer)

        Protected MustOverride Sub ParseBody(ByVal state As Object, ByVal binaryImage As Byte(), ByVal startIndex As Integer)

        Public ReadOnly Property Parent() As IChannelFrame Implements IChannelCell.Parent
            Get
                Return m_parent
            End Get
        End Property

        Public ReadOnly Property AlignOnDWordBoundry() As Boolean Implements IChannelCell.AlignOnDWordBoundry
            Get
                Return m_alignOnDWordBoundry
            End Get
        End Property

        Public Overrides ReadOnly Property BinaryLength() As Int16
            Get
                Dim length As Int16 = MyBase.BinaryLength

                If m_alignOnDWordBoundry Then
                    ' If requested, we align frame cells on 32-bit word boundries
                    Do Until length Mod 4 = 0
                        length += 1
                    Loop
                End If

                Return length

            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2005 TVA Code Library/Version/4.0/Source/TVA.PhasorProtocols/ChannelCellBase.cs
//*******************************************************************************************************
//  ChannelCellBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R. Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  3/7/2005 - James R. Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Security.Permissions;

namespace TVA.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent representation of any kind of data cell.
    /// </summary>
    /// <remarks>
    /// This phasor protocol implementation defines a "cell" as a portion of a frame, i.e., a logical unit of data.
    /// For example, a <see cref="DataCellBase"/> (dervied from <see cref="ChannelCellBase"/>) could be defined as a PMU
    /// within a frame of data, a <see cref="DataFrameBase"/>, that contains multiple PMU's coming from a PDC.
    /// </remarks>
    [Serializable()]
    public abstract class ChannelCellBase : ChannelBase, IChannelCell
    {
        #region [ Members ]

        // Fields
        private IChannelFrame m_parent;         // Reference to parent frame of this channel cell
        private ushort m_idCode;                // Numeric identifier of this logical unit of data (e.g., PMU ID code)

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="ChannelCellBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">A reference to the parent <see cref="IChannelFrame"/> for this <see cref="ChannelCellBase"/>.</param>
        /// <param name="idCode">The numeric ID code for this <see cref="ChannelCellBase"/>.</param>
        protected ChannelCellBase(IChannelFrame parent, ushort idCode)
        {
            m_parent = parent;
            m_idCode = idCode;
        }

        /// <summary>
        /// Creates a new <see cref="ChannelCellBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected ChannelCellBase(SerializationInfo info, StreamingContext context)
        {
            // Deserialize basic channel cell values
            m_parent = (IChannelFrame)info.GetValue("parent", typeof(IChannelFrame));
            m_idCode = info.GetUInt16("id");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets a reference to the parent <see cref="IChannelFrame"/> for this <see cref="ChannelCellBase"/>.
        /// </summary>
        public virtual IChannelFrame Parent
        {
            get
            {
                return m_parent;
            }
            set
            {
                m_parent = value;
            }
        }

        /// <summary>
        /// Gets or sets the parsing state for the this <see cref="ChannelCellBase"/>.
        /// </summary>
        new public virtual IChannelCellParsingState State
        {
            get
            {
                return base.State as IChannelCellParsingState;
            }
            set
            {
                base.State = value;
            }
        }

        /// <summary>
        /// Gets or sets the numeric ID code for this <see cref="ChannelCellBase"/>.
        /// </summary>
        /// <remarks>
        /// Most phasor measurement devices define some kind of numeric identifier (e.g., a hardware identifier coded into the device ROM); this is the
        /// abstract representation of this identifier.
        /// </remarks>
        public virtual ushort IDCode
        {
            get
            {
                return m_idCode;
            }
            set
            {
                m_idCode = value;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="ChannelCellBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("ID Code", IDCode.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            // Serialize basic channel cell values
            info.AddValue("parent", m_parent, typeof(IChannelFrame));
            info.AddValue("id", m_idCode);
        }

        #endregion
    }
}